File: csharp/GardenCatalogue-TestDesign-Kata/GardenCatalogueTest/PlantDataParserTests.cs
using System.Text;
using GardenCatalogue;
using NUnit.Framework;

namespace GardenCatalogueTest;

[TestFixture]
public class PlantDataParserTests
{
    private PlantDataParser _dataParser;

    [SetUp]
    public void SetUp()
    {
        BugConfigurations.Reset();
        _dataParser = new PlantDataParser();
    }

    [Test]
    public void Parse_EmptyStream_ReturnsEmptyList()
    {
        using var stream = new MemoryStream();
        var result = _dataParser.Parse(stream);
        Assert.That(result, Is.Empty);
    }

    [Test]
    public void Parse_SingleV1Plant_ReturnsCorrectData()
    {
        var plants = new List<Plant>
        {
            new PlantBuilder().WithName("Test Bush").WithType(PlantType.Bush).Build()
        };

        byte[] binaryData = CreateBinaryData(plants, version: 1);
        using var stream = new MemoryStream(binaryData);

        var result = _dataParser.Parse(stream).ToList();

        Assert.That(result.Count, Is.EqualTo(1));
        Assert.That(result[0].Name, Is.EqualTo("Test Bush"));
        Assert.That(result[0].Type, Is.EqualTo(PlantType.Bush));
    }

    [Test]
    public void Parse_ValidV1BinaryData_ReturnsPlants()
    {
        var plants = new List<Plant>
        {
            new PlantBuilder().WithName("Flower1").WithBloomPeriod(BloomPeriod.FromRange(Month.May, Month.July)).WithMaxHeight(0.5).WithSoil(SoilCondition.Sandy).WithLight(LightCondition.FullSun).Build(),
            new PlantBuilder().WithName("Bush1").WithType(PlantType.Bush).WithBloomPeriod(new BloomPeriod(Month.August)).WithMaxHeight(2.0).WithSoil(SoilCondition.Clay).WithLight(LightCondition.PartialShade).Build()
        };

        byte[] binaryData = CreateBinaryData(plants, version: 1);
        Console.WriteLine($"[DEBUG_LOG] Binary Data Length: {binaryData.Length}");
        using var stream = new MemoryStream(binaryData);

        var result = _dataParser.Parse(stream).ToList();

        Assert.That(result.Count, Is.EqualTo(2));
        Assert.That(result[0].Name, Is.EqualTo("Flower1"));
        Assert.That(result[0].Type, Is.EqualTo(PlantType.Flower), "First plant type should be Flower");
        Assert.That(result[0].MaxHeight, Is.EqualTo(0.5));
        Assert.That(result[0].Soil, Is.EqualTo(SoilCondition.Sandy));
        Assert.That(result[0].Light, Is.EqualTo(LightCondition.FullSun));
        Assert.That(result[0].BloomPeriod.Months, Is.EquivalentTo(new[] { Month.May, Month.June, Month.July }));
        
        Assert.That(result[1].Name, Is.EqualTo("Bush1"), "Second plant name mismatch");
        Assert.That(result[1].Type, Is.EqualTo(PlantType.Bush), "Second plant type should be Bush");
        Assert.That(result[1].MaxHeight, Is.EqualTo(2.0));
        Assert.That(result[1].Soil, Is.EqualTo(SoilCondition.Clay));
        Assert.That(result[1].Light, Is.EqualTo(LightCondition.PartialShade));
        Assert.That(result[1].BloomPeriod.Months, Is.EquivalentTo(new[] { Month.August }));
    }

    [Test]
    public void Parse_V2ExtendedData_ReturnsPlantsWithAllFields()
    {
        var plants = new List<Plant>
        {
            new PlantBuilder()
                .WithName("Lavender")
                .WithLatinName("Lavandula angustifolia")
                .WithArticleNumber("FLO-001")
                .WithBloomPeriod(BloomPeriod.FromRange(Month.June, Month.August))
                .WithMaxHeight(0.6)
                .WithSoil(SoilCondition.Sandy)
                .WithLight(LightCondition.FullSun)
                .WithProperty("Color", "Purple")
                .WithProperty("Fragrance", "High")
                .Build(),
             new PlantBuilder()
                .WithName("Minimal Plant")
                .Build()
        };

        byte[] binaryData = CreateBinaryData(plants, version: 2);
        using var stream = new MemoryStream(binaryData);

        var result = _dataParser.Parse(stream).ToList();

        Assert.That(result.Count, Is.EqualTo(2));
        var plant = result[0];
        Assert.That(plant.Name, Is.EqualTo("Lavender"));
        Assert.That(plant.LatinName, Is.EqualTo("Lavandula angustifolia"));
        Assert.That(plant.ArticleNumber, Is.EqualTo("FLO-001"));
        Assert.That(plant.Properties["Color"], Is.EqualTo("Purple"));
        Assert.That(plant.Properties["Fragrance"], Is.EqualTo("High"));

        var minimal = result[1];
        Assert.That(minimal.Name, Is.EqualTo("Minimal Plant"));
        Assert.That(minimal.LatinName, Is.Empty);
        Assert.That(minimal.ArticleNumber, Is.Empty);
        Assert.That(minimal.Properties, Is.Empty);
    }
    
    private static byte[] CreateBinaryData(IEnumerable<Plant> plants, int version = 2)
    {
        using var stream = new MemoryStream();
        using var writer = new BinaryWriter(stream, Encoding.UTF8);

        if (version == 2)
        {
            writer.Write("PLNT"u8.ToArray());
        }

        foreach (var plant in plants)
        {
            byte[] nameBytes = new byte[PlantDataParser.NameLength];
            byte[] actualNameBytes = Encoding.UTF8.GetBytes(plant.Name);
            Array.Copy(actualNameBytes, 0, nameBytes, 0, Math.Min(actualNameBytes.Length, PlantDataParser.NameLength));
            writer.Write(nameBytes);

            writer.Write((int)plant.Type);
            writer.Write(plant.MaxHeight);
            writer.Write((int)plant.Soil);
            writer.Write((int)plant.Light);

            int bloomMask = 0;
            foreach (var month in plant.BloomPeriod.Months)
            {
                bloomMask |= (1 << ((int)month - 1));
            }
            writer.Write(bloomMask);

            if (version == 2)
            {
                // LatinName (Length-prefixed)
                writer.Write(plant.LatinName);
                // ArticleNumber (Length-prefixed)
                writer.Write(plant.ArticleNumber);
                // Properties count
                writer.Write(plant.Properties.Count);
                foreach (var kvp in plant.Properties)
                {
                    writer.Write(kvp.Key);
                    writer.Write(kvp.Value);
                }
            }
        }

        return stream.ToArray();
    }
}

File: java/src/test/java/org/sammancoaching/PlantDataParserTests.java
package org.sammancoaching;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import static org.junit.jupiter.api.Assertions.*;

public class PlantDataParserTests {
    private PlantDataParser dataParser;

    @BeforeEach
    public void setup() {
        BugConfigurations.reset();
        dataParser = new PlantDataParser();
    }

    @Test
    public void testEmptyStreamEmptyList() throws IOException {
        ByteArrayInputStream stream = new ByteArrayInputStream(new byte[0]);
        List<Plant> result = dataParser.parse(stream);
        assertTrue(result.isEmpty());
    }

    @Test
    public void testV1SinglePlantValidResult() throws IOException {
        Plant plant = new Plant("Test Bush", "", "", PlantType.Bush, new BloomPeriod(List.of()), 0.0, SoilCondition.Any, LightCondition.FullSun, new HashMap<>());
        byte[] binaryData = createBinaryData(List.of(plant), 1);
        List<Plant> result = dataParser.parse(new ByteArrayInputStream(binaryData));

        assertEquals(1, result.size());
        assertEquals("Test Bush", result.get(0).name());
        assertEquals(PlantType.Bush, result.get(0).type());
    }

    @Test
    public void testV1TwoPlantsValidResult() throws IOException {
        Plant p1 = new Plant("Flower1", "", "", PlantType.Flower, BloomPeriod.fromRange(Month.May, Month.July), 0.5, SoilCondition.Sandy, LightCondition.FullSun, new HashMap<>());
        Plant p2 = new Plant("Bush1", "", "", PlantType.Bush, new BloomPeriod(List.of(Month.August)), 2.0, SoilCondition.Clay, LightCondition.PartialShade, new HashMap<>());
        
        byte[] binaryData = createBinaryData(List.of(p1, p2), 1);
        List<Plant> result = dataParser.parse(new ByteArrayInputStream(binaryData));

        assertEquals(2, result.size());
        assertEquals("Flower1", result.get(0).name());
        assertEquals(PlantType.Flower, result.get(0).type());
        assertEquals(0.5, result.get(0).maxHeight());
        assertEquals(SoilCondition.Sandy, result.get(0).soil());
        assertEquals(LightCondition.FullSun, result.get(0).light());
        assertEquals(List.of(Month.May, Month.June, Month.July), result.get(0).bloomPeriod().months());

        assertEquals("Bush1", result.get(1).name());
        assertEquals(PlantType.Bush, result.get(1).type());
        assertEquals(2.0, result.get(1).maxHeight());
        assertEquals(SoilCondition.Clay, result.get(1).soil());
        assertEquals(LightCondition.PartialShade, result.get(1).light());
        assertEquals(List.of(Month.August), result.get(1).bloomPeriod().months());
    }

    @Test
    public void testV2TwoPlantsValidResult() throws IOException {
        Plant p1 = new Plant("Lavender", "Lavandula angustifolia", "FLO-001", PlantType.Flower, BloomPeriod.fromRange(Month.June, Month.August), 0.6, SoilCondition.Sandy, LightCondition.FullSun, Map.of("Color", "Purple", "Fragrance", "High"));
        Plant p2 = new Plant("Minimal Plant", "", "", PlantType.Flower, new BloomPeriod(List.of()), 0.0, SoilCondition.Any, LightCondition.FullSun, Map.of());

        byte[] binaryData = createBinaryData(List.of(p1, p2), 2);
        List<Plant> result = dataParser.parse(new ByteArrayInputStream(binaryData));

        assertEquals(2, result.size());
        assertEquals("Lavender", result.get(0).name());
        assertEquals("Lavandula angustifolia", result.get(0).latinName());
        assertEquals("FLO-001", result.get(0).articleNumber());
        assertEquals("Purple", result.get(0).properties().get("Color"));
        assertEquals("High", result.get(0).properties().get("Fragrance"));

        assertEquals("Minimal Plant", result.get(1).name());
        assertEquals("", result.get(1).latinName());
        assertEquals("", result.get(1).articleNumber());
        assertTrue(result.get(1).properties().isEmpty());
    }

    private byte[] createBinaryData(List<Plant> plants, int version) throws IOException {
        ByteArrayOutputStream out = new ByteArrayOutputStream();
        if (version == 2) {
            out.write("PLNT".getBytes(StandardCharsets.UTF_8));
        }

        for (Plant plant : plants) {
            byte[] nameBytes = new byte[32];
            byte[] actualName = plant.name().getBytes(StandardCharsets.UTF_8);
            System.arraycopy(actualName, 0, nameBytes, 0, Math.min(actualName.length, 32));
            out.write(nameBytes);

            ByteBuffer bb = ByteBuffer.allocate(24).order(ByteOrder.LITTLE_ENDIAN);
            bb.putInt(plant.type().ordinal());
            bb.putDouble(plant.maxHeight());
            bb.putInt(plant.soil().ordinal());
            bb.putInt(plant.light().ordinal());

            int bloomMask = 0;
            for (Month m : plant.bloomPeriod().months()) {
                bloomMask |= (1 << (m.getValue() - 1));
            }
            bb.putInt(bloomMask);
            out.write(bb.array());

            if (version == 2) {
                writeString(out, plant.latinName());
                writeString(out, plant.articleNumber());
                
                ByteBuffer propCountBuf = ByteBuffer.allocate(4).order(ByteOrder.LITTLE_ENDIAN);
                propCountBuf.putInt(plant.properties().size());
                out.write(propCountBuf.array());

                for (Map.Entry<String, String> entry : plant.properties().entrySet()) {
                    writeString(out, entry.getKey());
                    writeString(out, entry.getValue());
                }
            }
        }
        return out.toByteArray();
    }

    private void writeString(ByteArrayOutputStream out, String s) throws IOException {
        byte[] data = s.getBytes(StandardCharsets.UTF_8);
        write7BitEncodedInt(out, data.length);
        out.write(data);
    }

    private void write7BitEncodedInt(ByteArrayOutputStream out, int value) {
        while (value >= 0x80) {
            out.write((value & 0x7f) | 0x80);
            value >>= 7;
        }
        out.write(value);
    }
}
